libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+  g_autoptr(AppendStoresTaskData) data = append_stores_task_data_new (ka_proxy,
+                                                                      proxy_func,
+                                                                      proxy_return_destroy,
+                                                                      proxy_func_data);
+
+  g_task_set_return_on_cancel (task, TRUE);
+  g_task_set_task_data (task,
+                        g_steal_pointer (&data),
+                        (GDestroyNotify) append_stores_task_data_free);
+  g_task_run_in_thread (task, append_stores_task_from_proxy_thread);
+}
+
+typedef gpointer (*ElementCopyFunc) (gpointer element);
+
+static GPtrArray *
+flat_map_ptr_arrays (GPtrArray       *ptr_arrays,
+                     ElementCopyFunc  element_copy_func,
+                     GDestroyNotify   element_free_func)
+{
+  g_autoptr(GPtrArray) concatenated = NULL;
+  guint len = 0;
+  guint i = 0;
+  guint j = 0;
+
+  /* O(N) count up of all pointer array lengths */
+  for (i = 0; i < ptr_arrays->len; ++i)
+    {
+      GPtrArray *ptr_array = g_ptr_array_index (ptr_arrays, i);
+      len += ptr_array->len;
+    }
+
+  concatenated = g_ptr_array_new_full (len, element_free_func);
+
+  /* O(N) copying of elements */
+  for (i = 0; i < ptr_arrays->len; ++i)
+    {
+      GPtrArray *ptr_array = g_ptr_array_index (ptr_arrays, i);
+
+      for (j = 0; j < ptr_array->len; ++j)
+        g_ptr_array_add (concatenated,
+                         element_copy_func (g_ptr_array_index (ptr_array, j)));
+    }
+
+  return g_steal_pointer (&concatenated);
+}
+
+static void
+marshal_word_quote_into_store (GObject      *source,
+                               GAsyncResult *result,
+                               gpointer      user_data)
+{
+  g_autoptr(GTask) task = user_data;
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GPtrArray) word_quote_results = g_task_propagate_pointer (G_TASK (result),
+                                                                      &local_error);
+  g_autoptr(GPtrArray) word_quote_card_results = g_ptr_array_new_with_free_func (g_object_unref);
+  EosDiscoveryFeedWordCardStore *word_store = NULL;
+  EosDiscoveryFeedQuoteCardStore *quote_store = NULL;
+  EosDiscoveryFeedWordQuoteCardStore *store = NULL;
+
+
+  if (word_quote_results == NULL)
+    {
+      g_task_return_error (task, g_steal_pointer (&local_error));
+      return;
+    }
+
+  /* Unpack the word/quote results and concatenate the pointer arrays */
+  if (word_quote_results->len != 2)
+    {
+      g_task_return_error (task,
+                           g_error_new (G_IO_ERROR,
+                                        G_IO_ERROR_FAILED,
+                                        "Expected exactly two results for word/quote query."));
+      return;
+    }
+
+  word_store = g_task_propagate_pointer (G_TASK (g_ptr_array_index (word_quote_results, 0)),
+                                         &local_error);
```

Comment:
Are you sure that the order of results is guaranteed here?

Reply:
Hmm, that's a good point. Since we query both concurrently I suppose it is possible that you could get them out of order. Let me double check how this was implemented.

Reply:
Ah, yes, the order is guaranteed, since we write the result into the allocated result slot (and we allocate the `word` result slot first: https://github.com/endlessm/eos-discovery-feed/pull/92/files#diff-9822bca115b6cda571d6578fb4d8d4a7R69